design/client-policies.md

```diff
@@ -0,0 +1,191 @@
+
+## Use-cases
+
+Client Policies realize the following three points mentioned in Motivataion as follows.
+
+### 1. Setting policies on what configuration a client can have
+
+The current keycloak has already supported them as [Client Registration Policies](https://www.keycloak.org/docs/latest/securing_apps/index.html#client-registration-policies).
+
+However, it can only cover OIDC Dynamic Client Registration.
+
+Client Policies cover not only what Client Registration Policies can do but other client registration and configuration ways.
+
+In the future, Client Registration will be replaced by Client Policies.
+
+### 2. Validation of client configurations
+
+The current keycloak has already supported validation on whether the client follows settings like Proof Key for Code Exchange, Request Object Signing Algorithm, Holder-of-Key Token, etc on some endpoints like Authorization Endpoint, Token Endpoint, etc.
+
+These can be specified by each setting item (on Admin Console, switch, pulldown menu, etc). To make the client application secure, user need to set a lot of such settings in appropriate way, which makes it difficult for the user to secure their client application.
+
+Client Policies can do these validation of client configurations mentioned just above. In the futre, these each setting items will be replaced by Client Policies. 
+
+The pre-set client policies for following some security profiles (FAPI, SPA, Native App, etc) are provided, which makes it easy for the user to secure their client application.
+
+### 3. Conformance to a profile such as FAPI
+
+It is treated in [Client Conformance Profiles](https://issues.redhat.com/browse/KEYCLOAK-11612) based on this Client Policies. Which kind of security profiles are supported and how to realize them are discussed on 
+[Client Conformance Profiles](https://issues.redhat.com/browse/KEYCLOAK-11612).
+
+## Implementation details
+
+Client Policies consists of the three building blocks, Policy, Condition, and Executor.
+
+### Condition : which client
+
+A condition determins to which client a policy is adopted. The condition checks whether one specified criteria is satisfied. Such the task is that checking whether the [access type](https://www.keycloak.org/docs/latest/server_admin/index.html#oidc-clients) of the client is confidential.
+
+#### How to implement
+
+A condition can be implemented as a provider (`ClientPolicyConditionSpi`, `ClientPolicyConditionFactory`, `ClientPolicyCondition`). Developers can add their own condition by implementing this provider.
+
+#### How to configure
+
+A condition can be configuable the same as other configurable provider. What can be configured depends on each condition's nature.
+
+#### Which kind of conditions are provided
+
+The following conditions are provided:
+
+* The way of creating/updating a client (OIDC Dynamic Client Registration - Authenticated, Anonymous, Admin REST API - Admin Console, etc.)
```

Comment:
Should probably also have some conditions on the author of the client. User/group/role of creator of the client, or different policies based on different initial access tokens, etc.

Reply:
IMO, broadly speaking, keycloak user can register/update a client by 2 way.

1. Admin REST API (used by Admin Console, etc.)
2. Dynamic Client Registration (Anonymous/Authenticated)

As for the former, the creator of the client can be specified by 

1. The user who has the authority to creating a client
2. The client (e.g. security-admin-console)

We should prepare conditions based on properties of them (e.g. user, group, role).

As for the latter, we should prepare conditions checking whether Anonymous or Authenticated registration/update.

As for Authenticated registration(with initial access token) and update(with registration access token), we should prepare conditions based on the contents of their tokens (e.g. some claim included in the initial access token). However, we need to determine which claims are used for conditions.

WDYT?

Reply:
Dynamic client registration can also be invoked with a regular token, so probably look at it from who, not what endpoint. We have:

* Anonymous
* User - role, group, etc.
* Initial access token - Not sure what's best here, but would be good to allow creating different tokens with different access. Wonder if the ideal would just be to allow assigning for a given user/service-account or something like that? Then we're back to anonymous and user to check what policies to apply.

Reply:
On dynamic client registration, I think "User" stands for Service Account because the client application itself request client registration with some kind of token.
On registration by Admin REST API, I think "User" stands for the end user like keycloak's administrator. Is that right?

As for restriction based on contents of the token used on dynamic client registration, I think it might be sufficient to use the existing properties (role, scope, etc.) .

Reply:
For dynamic client registration it can also be a regular user. You may for example have some developer portal where a user can login to generate an access token that allows them to register clients.

For admin REST API, yes it would have to be an admin user.

Reply:
The ideal would be if we have a single way for admin, initial access, and user, to somehow say apply this policy to this group of users, or this policy to this group of users. That does suggest that potentially an initial access token should be issued to a specific user.

Reply:
I see, on dynamic client registration, "User" stands for the end user who was authenticated to get an access token for generating a new client, not Service Account of the existing client that actually accesses the registration endpoint with the access token.